### ISSUE TITLE
small object detection

### ISSUE THREAD

**Issue by RoshiniEvangelin** [ultralytics/yolov5#13657]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi, I trained YOLOv5 on my custom dataset with two classes (defect, background).
After training, the confusion matrix only shows predictions in the background column, and all true defect samples are being classified as background.
Dataset size: (e.g., 1000 defect, 5000 background)

Epochs: 999
Batch size: 16
Image size: 1028

Problem:
The confusion matrix shows 100% predictions as background.
Even true defect objects are misclassified as background

### Additional

<img width="3000" height="2250" alt="Image" src="https://github.com/user-attachments/assets/77889009-d713-4604-aa1f-b4dded02ef1c" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @RoshiniEvangelin, thank you for your interest in YOLOv5 🚀! This is an automated response to help you get unblocked quickly, and an Ultralytics engineer will also assist you here soon. For general guidance, please visit our ⭐️ [Tutorials](https://docs.ultralytics.com/yolov5/) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data/) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution/).
If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it.
If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). Screenshots of the confusion matrix and a few representative images are very helpful for small object detection questions 👍
## Requirements
[**Python>=3.8.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/). To get started:
```bash
git clone https://github.com/ultralytics/yolov5  # clone
cd yolov5
pip install -r requirements.txt  # install
```
## Environments
YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://bit.ly/yolov5-paperspace-notebook"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"></a> <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml"><img src="https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg" alt="YOLOv5 CI"></a>
If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by RoshiniEvangelin** [comment]
this is my model yaml file 
# YOLOv5 small object detection config
# Save as yolov5s_smallobj.yaml

nc: 80  # number of classes (update to your dataset)
depth_multiple: 0.33
width_multiple: 0.50

anchors:
  - [5, 6, 8, 14, 15, 11]          # P2/4 (tiny)
  - [10, 13, 16, 30, 33, 23]       # P3/8 (small)
  - [30, 61, 62, 45, 59, 119]      # P4/16 (medium)
  - [116, 90, 156, 198, 373, 326]  # P5/32 (large)

backbone:
  [
    [-1, 1, Conv, [64, 6, 2, 2]],     # 0-P1/2
    [-1, 1, Conv, [128, 3, 2]],       # 1-P2/4
    [-1, 3, C3, [128]],               # 2
    [-1, 1, Conv, [256, 3, 2]],       # 3-P3/8
    [-1, 6, C3, [256]],               # 4
    [-1, 1, Conv, [512, 3, 2]],       # 5-P4/16
    [-1, 9, C3, [512]],               # 6
    [-1, 1, Conv, [1024, 3, 2]],      # 7-P5/32
    [-1, 3, C3, [1024]],              # 8
    [-1, 1, SPPF, [1024, 5]],         # 9
  ]

head:
  [
    [-1, 1, Conv, [512, 1, 1]],                     # 10
    [-1, 1, nn.Upsample, [None, 2, "nearest"]],     # 11
    [[-1, 6], 1, Concat, [1]],                      # 12
    [-1, 3, C3, [512, False]],                      # 13

    [-1, 1, Conv, [256, 1, 1]],                     # 14
    [-1, 1, nn.Upsample, [None, 2, "nearest"]],     # 15
    [[-1, 4], 1, Concat, [1]],                      # 16
    [-1, 3, C3, [256, False]],                      # 17 (P3/8)

    [-1, 1, Conv, [128, 1, 1]],                     # 18 ↓
    [-1, 1, nn.Upsample, [None, 2, "nearest"]],     # 19
    [[-1, 2], 1, Concat, [1]],                      # 20 ← concat with backbone P2
    [-1, 3, C3, [128, False]],                      # 21 (P2/4)

    [-1, 1, Conv, [256, 3, 2]],                     # 22
    [[-1, 17], 1, Concat, [1]],                     # 23 (fix concat: match P3 path)
    [-1, 3, C3, [256, False]],                      # 24 (P3/8 downsampled again → aligns with P4)

    [-1, 1, Conv, [512, 3, 2]],                     # 25
    [[-1, 13], 1, Concat, [1]],                     # 26 (fix concat: match P4 path)
    [-1, 3, C3, [512, False]],                      # 27 (P4/16 downsampled again → aligns with P5)

    [[21, 17, 24, 27], 1, Detect, [nc, anchors]],   # Detect(P2, P3, P4, P5)
  ]

**Comment by pderrenger** [comment]
Thanks for the details and the YAML. The “background” column in the confusion matrix means your ground-truth defects weren’t detected; for object detection you should not create a “background” class—set your data.yaml to nc: 1 with names: [defect], ensure label files contain only class 0 boxes, and retrain. Also use an img size that’s a multiple of 32 (e.g., 1024 or 1280, not 1028) and start from a stock model/anchors before custom P2 changes: python train.py --img 1280 --batch 16 --epochs 100 --data data.yaml --weights yolov5s.pt, then verify at low conf: python val.py --weights runs/train/exp/weights/best.pt --data data.yaml --img 1280 --conf 0.001 --plots. To quickly sanity-check labels, run a 1‑epoch dry run and inspect train_batch*.jpg: python train.py --weights '' --img 1024 --batch 16 --epochs 1 --data data.yaml --nosave --cache. See the Confusion Matrix guide for how the background column is interpreted: Confusion Matrix Explained | Ultralytics.
